adding dra docs (#162)
NVIDIA/cloud-native-docs: gpu-operator/dra-driver.rst

###################################
NVIDIA GPU DRA Driver
###################################

*******************************
About the NVIDIA GPU DRA Driver
*******************************

> The NVIDIA GPU DRA Driver leverages the Kubernetes Dynamic Resource Allocation (DRA) API to support NVIDIA IMEX channels available in GH200 and GB200 GPUs.

Reviewer: May be we should add a bit about imex channels. E.g. 
.... to support NVIDIA IMEX (Internode Memory Exchange/Management Service) channels available in GH200 and G200 systems that allows the GPUs to directly read / write each other’s memory over a high-bandwidth NVLink.

Reviewer 2: defined in line 13 and 14

The NVIDIA GPU DRA Driver creates and manages IMEX channels through the creation of a ComputeDomain custom resource. Use this custom resource to define your resource templates, and then reference the templates within your workload specs.

An IMEX channel is a construct that allows a set of GPUs to directly read and write each other's memory over a high-bandwidth NVLink. 

> The NVLink connection may either be directly between GPUs on the same node or between GPUs on separate nodes connected by an NVSwitch. 

Reviewer: oh you defined it here.

Once an IMEX channel has been established for a set of GPUs, they are free to read and write each other's memory via extensions to the CUDA memory call APIs.

> The ability to support IMEX channels on GH200 and GB200 systems is essential, as they have been designed specifically to exploit the use of IMEX channels to turn a rack of GPU machines (each with a small number of GPUs) into a giant supercomputer with up to 72 GPUs communicating at full NVLink bandwidth. 

Reviewer: I think it is imp to first summarize what is to follow. That imex channels are meant for multi-node communication

IMEX channel (by definition) is a resource that spans multiple nodes, hence the ability to support IMEX channels on GH200 and GB200 systems is essential. These GH200 and GB200 systems are designed  specifically to leverage the use of IMEX channels to turn a rack of GPU machines (each with a small number of GPUs) into a giant supercomputer with up to 72 GPUs communicating at full NVLink bandwidth.

This allows you to get the most use out of your available GPUs without any additional latency burdens.  

Kubernetes Dynamic Resource Allocation (DRA), available as beta in Kubernetes v1.32, is an API for requesting and sharing resources between pods and containers inside a pod. 
This feature treats specialized hardware as a definable and reusable object and provides the necessary primitives to support cross-node resources such as IMEX channels. 

> Along with the NVIDIA GPU DRA Driver, you are able to use DRA to define IMEX channel resources that can be managed by Kubernetes.

Reviewer: NVIDIA GPU DRA Driver uses DRA features to define IMEX channel resources that can be managed by Kubernetes.

Refer to the `Kubernetes DRA documentation <https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/>`__ for details on this feature. 

*************
Prerequisites
*************

- GH200 and GB200 GPUs with Mulit-Node NVLink connections between GPUs.

- 565 or later driver installed on all GPU nodes. 
  Drivers must be pre-installed on all GPU nodes before installing the NVIDIA GPU Operator as operator managed drivers are not supported at this time.

- IMEX packages installed on GPU nodes with systemd service disabled.

>   The IMEX package verisons must match the installed driver version.

Reviewer: Refer to the release notes for the exact version of the driver and IMEX to install.

Author: @guptaNswati can you share the a link to release notes you are referring to? thanks!

  The following example shows how to install IMEX drivers. 

  .. code-block:: console 

> 

Reviewer: These are assuming, you already installed NVIDIA drivers

DRIVER_VERSION=$(nvidia-smi -i 0 --query-gpu=driver_version --format=csv,noheader,nounits)
IMEX_DRIVER_VERSION="${DRIVER_VERSION%%.*}"
TARGETARCH=$(uname -m | sed -E 's/^x86_64$/amd64/; s/^(aarch64|arm64)$/arm64/')

Reviewer 2: yes, line 32 defines the second prerequisite for installing drivers

      $ curl -fsSL -o /tmp/nvidia-imex-${IMEX_DRIVER_VERSION}_${DRIVER_VERSION}-1_${TARGETARCH}.deb https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/sbsa/nvidia-imex-${IMEX_DRIVER_VERSION}_${DRIVER_VERSION}-1_${TARGETARCH}.deb && dpkg -i /tmp/nvidia-imex-${IMEX_DRIVER_VERSION}_${DRIVER_VERSION}-1_${TARGETARCH}.deb && \
      nvidia-imex --version && \

>       ls /etc/nvidia-imex && \

Reviewer: dont need this ls

      rm -rf /tmp/nvidia-imex_${IMEX_DRIVER_VERSION}_${DRIVER_VERSION}-1_${TARGETARCH}.deb

  Then disable systemd service.

  .. code-block:: console 

    $ systemctl disable --now nvidia-imex.service && systemctl mask nvidia-imex.service

> - Kubernetes v1.32 multi-node cluster with the DynaticResourceAllocation feature gate enabled. 

Reviewer 3: Typo:  "Dynatic"

Author: updated, thanks!

  The following is a sample for enabling DRA feature gates. 
  Refer to the Kubernetes documentation for full details on `enabling DRA on your cluster <https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/#enabling-dynamic-resource-allocation>`__.

>   .. literalinclude:: ./manifests/input/kubeadm-init-config.yaml

Reviewer: are you going to add the yaml?

If a subset of the nodes associated with a ComputeDomain are capable of communicating over IMEX, the NVIDIA Kubernetes DRA will set up a one-off IMEX domain to allow GPUs to communicate over their multi-node NVLink connections. Multiple such IMEX domains will be created as necessary depending on the number (and availability) of nodes allocated to the ComputeDomain. 

.. literalinclude:: ./manifests/input/dra-compute-domain-crd.yaml
   :language: yaml
   :caption: Sample NVIDIA DRA Driver ComputeDomain Custom Resource Manifest

The following table describes some of the fields in the custom resource.

.. list-table::
   :header-rows: 1
   :widths: 20 60 20

   * - Field
     - Description
     - Default Value

   * - ``channel.resourceClaimTemplate.name`` (required)
     - Specifies the ``name`` of the ResourceClaimTemplate to create.
     - None

   * - ``numNodes`` (required)
     - Specifies the number of nodes in the IMEX domain.
     - None

Node and Pod Affitnity Strategies
=========================================

> The ComputeDomain object is not tied directly to the notion of an IMEX channel. 

Reviewer: I would rephrase it:

A ComputeDomain isn’t strictly about IMEX channels—it’s about running workloads across a group of compute nodes. This means even if some nodes are not IMEX capable, they can still be part of the same ComputeDomain. To control where your workloads run, you should use NodeAffinity and PodAffinity rules.

As the workload ``imex-channel-injection-pod`` referencing the ResourceClaimTemplate for workloads, ``imex-channel-0``, get scheduled, they trigger the DRA Driver to request access to the same IMEX channel on whatever node they land on. 

Once scheduled to a node, the DRA driver adds a Node label for the ComputeDomain to the node where the workload has been scheduled to indicate the node is part of that ComputeDomain.
This label is used as a NodeSelector on the DaemonSet mentioned above to trigger the scheduling of its pods to specific nodes.
Once all workloads running in a ComputeDomain have run to completion, the label gets removed (even if the ComputeDomain itself hasn't been deleted yet).
This allows these nodes to be reused for other ComputeDomains and blocks the workloads from running until the IMEX daemon on that node has been started.

The addition of this Node label triggers the DaemonSet to deploy an IMEX daemon to that node and start running it.

Once all daemons have been fully started, the DRA driver unblocks each worker, injects its IMEX channel into the worker and allows it to start running.

> View the Compute Domain resources on your cluster

Reviewer: we should also add the MPI nvbandwidth example.

Reviewer: Its also described here https://github.com/NVIDIA/k8s-dra-driver-gpu/discussions/249

Author: will add that!

        Channel:
          Resource Claim Template:
            Name:   imex-channel-0
        Num Nodes:  1
      Events:       <none>

#. View IMEX channel creation logs.

   .. code-block:: console

      $ kubectl logs -n nvidia-dra-driver-gpu -l resource.nvidia.com/computeDomain --tail=-1

   *Example Output*

    .. literalinclude:: ./manifests/output/imex-logs.txt
        :language: yaml
  
Node and Pod Affitnity Strategies
=================================

A ComputeDomain isn't strictly about IMEX channels—it's about running workloads across a group of compute nodes. 
This means even if some nodes are not IMEX capable, they can still be part of the same ComputeDomain.
To control where your workloads run, you should use NodeAffinity and PodAffinity rules.

For example you could set PodAffinity with a preferred topologyKey set to ``nvidia.com/gpu.clique`` for workloads to span multiple NVLink domains but want them packed as tightly as possible. Or use a required topologyKey set to ``nvidia.con/gpu.clique`` when you require all workloads deployed into the same NVLink domain, but don't care which one. 

******************************************************************

> Run a Multi-node nvbandwidth Test Requiring IMEX Channels with MPI

Author: i need to do some more testing to clean up this example and add some explanations around whats happening.

.. license-header
  SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  SPDX-License-Identifier: Apache-2.0

############################################################

> Install the NVIDIA GPU DRA Driver and Configure IMEX Support

Reviewer 4: I think what we're supporting from a customer perspective is "Multi-Node NVLink" and *how* we're supporting it is through IMEX channels. The IMEX channels are an implementation detail from the user perspective.

I also believe that the DRA driver is no longer called the IMEX driver, but the `Compute Domain DRA Driver`.

Reviewer 5: Quick feedback on naming.

I am OK with "NVIDIA GPU DRA Driver".

We need consistency. Naming is hard.

Here are my preferences:

> NVIDIA DRA Driver for GPUs

> Kubernetes DRA driver for NVIDIA GPUs

The latter is most precise and intuitive I think. But maybe too long.

I trust that a certain decision has already been made, and we are late in the game with naming discussion. Yet, it's an important discussion.

For the Helm chart (public) listing I picked "NVIDIA DRA Driver for GPUs", see

https://catalog.ngc.nvidia.com/orgs/nvidia/helm-charts/nvidia-dra-driver-gpu

Here is how we should think about it:

> The DRA driver for NVIDIA GPUs enables multi-node GPU workloads in Kubernetes environments

That's the very high-level.

I agree with Evan that anything IMEX is an (important) implementation detail.

Reviewer 5: > I also believe that the DRA driver is no longer called the IMEX driver

And yes yes yes to this. We should think that it never was named IMEX driver. This was an organizational brainfart. :)

Reviewer 5: After Slack discussion: for now, we decided to use "NVIDIA DRA Driver for GPUs", like we did for nSpect and also the Helm chart registry. Kevin also prefers this:

> I usually prefer NVIDIA DRA Driver for GPUs so there’s not 3 all caps acronyms right next to each other

Author: naming should be updated in the docs

############################################################

> THe NVIDIA GPU DRA Driver is an additional component you can install after the GPU Operator that enables you to use the Kubernetes DRA feature to define IMEX channel resources that are managed by Kubernetes.

Reviewer 4: What about:
```suggestion
The NVIDIA GPU DRA Driver is an additional component you can install alongside the GPU Operator that enables you to use the Kubernetes DRA feature to support Multi-Node NVLink in distributed applications.

```
We may also want to link to the Kubernetes DRA documentation? https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/

Author: updated

This page details more information about the GPU DRA Driver, including how to install it and examples of deploying workloads using IMEX channels.

*******************************

> About the NVIDIA GPU DRA Driver

Reviewer 4: What is the "GPU DRA Driver" in this context? Is it the driver that is managing the lifecycle of compute domains or access to GPUs themselves?

Reviewer 5: > What is the "GPU DRA Driver" in this context? 

The "NVIDIA GPU DRA Driver" in this document always refers to the component developed in https://github.com/NVIDIA/k8s-dra-driver-gpu

> Is it the driver that is managing the lifecycle of compute domains 

in that sense: yes

*******************************

> The NVIDIA GPU DRA Driver leverages the Kubernetes Dynamic Resource Allocation (DRA) API to support NVIDIA IMEX channels available in GH200 and GB200 GPUs.

Reviewer 4: Does something like the following better capture the intent of the COMPUTE DOMAIN DRIVER:

> The NVIDIA GPU DRA Driver provides a Compute Domain abstraction (as a Kubernetes CRD) that allows distributed applications to make use of technologies such as Mulit-node NVLink if available. The underlying "connectivity" (not sure about this word) is managed by the NVIDIA GPU DRA Driver to ensure portability workloads.

Reviewer 4: Note that this does not mention IMEX channels.

The NVIDIA GPU DRA Driver creates and manages IMEX channels through the creation of a ComputeDomain custom resource. You use this custom resource to define your resource templates, and then reference the templates within your workload specs. The GPU DRA Driver will handle creating the cooresponding IMEX channels for you. 

An IMEX channel is a construct that allows a set of GPUs to directly read and write each other's memory over a high-bandwidth NVLink. 
The NVLink connection may either be directly between GPUs on the same node or between GPUs on separate nodes connected by an NVSwitch. 

> Once an IMEX channel has been established for a set of GPUs, they are free to read and write each other's memory via extensions to the CUDA memory call APIs.

Reviewer 5: My hunch is that we should not go into this detail here but just link to reference documentation.

Reviewer 5: Reference docs are here, for example: https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MALLOC__ASYNC.html#group__CUDA__MALLOC__ASYNC_1g8aa4c143dbc20293659cd883232b95f2

There is the beautiful statement of

> When exporter and importer CUDA processes have been granted access to the same IMEX channel, they can securely share memory.

And that's already a better description because of the emphasis on _securely_.

Maybe we can borrow that statement, and otherwise just link to refdocs.

An IMEX channel is a resource that spans multiple nodes. 
GH200 and GB200 systems are designed specifically to leverage the use of IMEX channels to turn a rack of GPU machines, each with a small number of GPUs, into a giant supercomputer with up to 72 GPUs communicating at full NVLink bandwidth.
This allows you to get the most use out of your available GPUs without any additional latency burdens.  

Kubernetes DRA, available as beta from Kubernetes v1.32, is an API for requesting and sharing resources between pods and containers inside a pod. 
This feature treats specialized hardware as a definable and reusable object and provides the necessary primitives to support cross-node resources such as IMEX channels. 
NVIDIA GPU DRA Driver uses DRA features to define IMEX channel resources that can be managed by Kubernetes.

Refer to the `Kubernetes DRA documentation <https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/>`__ for details on this feature. 

*********************************
Install the NVIDIA GPU DRA Driver 
*********************************

> The GPU DRA Driver is an addiitonal component that can be installed after you've installed the GPU Operator on your clsuter.

Reviewer 4: I don't think it's strictly speaking *required* that the GPU Operator be installed first, but may make things simpler. 

Also a typo:
```suggestion
The GPU DRA Driver is an addiitonal component that can be installed after you've installed the GPU Operator on your k8s cluster.

```

Prerequisites
=============

> - GH200 and GB200 GPUs with Mulit-Node NVLink connections between GPUs.

Reviewer 4: Note that the Compute Domain driver does not *require* a GH200 or GB200 system. The driver will function on systems that do not have MNNVL.

> - GH200 and GB200 GPUs with Mulit-Node NVLink connections between GPUs.

Author: @elezar what do you mean by function? like it will deploy and not throw errors, or will it actually create resources on the cluster, even though there is no MNNVL

> - GH200 and GB200 GPUs with Mulit-Node NVLink connections between GPUs.

Author: If i'm understanding this correctly, it might function, but we are only supporting MNNVL for this release. we can adjust the wording and split up the requirements section further as we add support for other use cases.

- Kubernetes v1.32 multi-node cluster with the DynamitcResourceAllocation feature gate enabled. 

>   The following is a sample for enabling DRA feature gates. 

Reviewer 4: ```suggestion
  The following is a sample for enabling the required feature gates and API groups. 

```

Author: updated.

  Refer to the Kubernetes documentation for full details on `enabling DRA on your cluster <https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/#enabling-dynamic-resource-allocation>`__.

  .. literalinclude:: ./manifests/input/kubeadm-init-config.yaml
    :language: yaml
    :caption: Sample Kubeadm Init Config with DRA Feature Gates Enabled

> - The NVIDIA GPU Operator v25.3.0 or later installed with CDI enabled on all nodes and NVIDIA GPU Driver 565 or later.

Reviewer 4: The v565 driver is only for GH200 / GB200 systems where MNNVL is supported. For other systems older driver versions should be sufficient.

Reviewer 6: This entire page is specific to documenting MNNVL. So I think it is okay to mention the minimum driver version that supports MNNVL.

  
  A sample Helm install command below includes enabling CDI with ``cdi.enabled=true``.
  Refer to the install documentation for details on `enabling CDI <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/getting-started.html#common-chart-customization-options>`__.

  .. code-block:: console

      $ helm install --wait --generate-name \
              -n gpu-operator --create-namespace \
              nvidia/gpu-operator \
              --version=${version} \
              --set cdi.enabled=true

>   Note if you want to install the GPU DRA Driver using pre-installed drivers, you must install NVIDIA GPU Driver 565 or later, the corresponding IMEX packages on GPU nodes, and disable the IMEX systemd service before installing the GPU Operator.

Reviewer 4: ```suggestion
  Note if you want to install the NVIDIA GPU DRA Driver using pre-installed drivers, you must install NVIDIA GPU Driver 565 or later, the corresponding IMEX packages on GPU nodes, and disable the IMEX systemd service before installing the GPU Operator.

```

  Refer to the documentation on `installing the GPU Operator with pre-installed drivers <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/getting-started.html#pre-installed-nvidia-gpu-drivers>`__ for more details.

Install the GPU DRA Driver with Helm
====================================

#. Add the NVIDIA Helm repository:

   .. code-block:: console

      $ helm repo add nvidia https://helm.ngc.nvidia.com/nvidia \
          && helm repo update

#. Install NVIDIA DRA Driver:

   .. code-block:: console

      $ helm install nvidia-dra-driver-gpu nvidia/nvidia-dra-driver-gpu \
            --version="25.3.0-rc.1" \
            --create-namespace \
            --namespace nvidia-dra-driver-gpu \
            --set nvidiaDriverRoot=/run/nvidia/driver \
            --set nvidiaCtkPath=/usr/local/nvidia/toolkit/nvidia-ctk \

>             --set resources.gpus.enabled=false

Reviewer 5: Thank you for following this :heart:

Common Chart Customization Options
==================================

The following options are available when using the Helm chart.
These options can be used with ``--set`` when installing with Helm.

The following table identifies the most frequently used options.
To view all the options, run ``helm show values nvidia/nvidia-dra-driver-gpu``.

.. list-table::
   :widths: 20 50 30
   :header-rows: 1

   * - Parameter
     - Description
     - Default

   * - ``nvidiaDriverRoot``
     - Specifies the driver root on the host.

>        For Operator managed drivers, use ``/run/nvidia/driver``.

Reviewer 4: ```suggestion
       For Operator-managed drivers, use ``/run/nvidia/driver``.

```

>        For pre-installed drivers, use ``/``.

Reviewer 5: Can we provide a _recommended_ mode of operation? 

Something like "Use Operator-managed drivers if in doubt".

     - ``/``

   * - ``nvidiaCtkPath``

>      - Specifies the path of The NVIDIA Container Tool Kit binary (nvidia-ctk) on the host, as it should appear in the the generated CDI specification.

Reviewer 4: The "generated CDI specification" is probably an implementation detail that should not be relevant in this case.

>        The exact path depends on the system that runs on the node.

Reviewer 5: Maybe here we should also say sth like  "use the Operator-managed nvidia-ctk if in doubt" 

CC @elezar

Reviewer 4: I think we can do something like we do above for the `nvidiaDriverRoot` and call out the two options with their default values.

- `/usr/bin/nvidia-ctk` for a pre-installed NVIDIA Container Toolkit
- `/usr/local/nvidia/toolkit/nvidia-ctk` for an GPU Operator-installed NVIDIA Container Toolkit

      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-7kdg9          1/1     Running   0          7m27s
      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-bd6gn          1/1     Running   0          7m27s
      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-bzm6p          1/1     Running   0          7m26s
      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-xjm4p          1/1     Running   0          7m27s

#. Confirm that all GPU nodes are labeled with clique ids.

   .. code-block:: console

      $ (echo -e "NODE\tLABEL\tCLIQUE"; kubectl get nodes -o json | \
        jq -r '.items[] | [.metadata.name, "nvidia.com/gpu.clique", .metadata.labels["nvidia.com/gpu.clique"]] | @tsv') | \
        column -t

   *Example Output*

   .. code-block:: output
    
      NODE                  LABEL                  CLIQUE
      node1                 nvidia.com/gpu.clique  1fbed3a8-bd74-4c83-afcb-cfb75ebc9304.1
      node2                 nvidia.com/gpu.clique  1fbed3a8-bd74-4c83-afcb-cfb75ebc9304.1
      node3                 nvidia.com/gpu.clique  1fbed3a8-bd74-4c83-afcb-cfb75ebc9304.1
      node4                 nvidia.com/gpu.clique  1fbed3a8-bd74-4c83-afcb-cfb75ebc9304.1

> The GPU DRA Driver adds a Clique ID to each GPU node. 

Author: updated

This is an unique identifier within an NVLink domain (physically connected GPUs over NVLink) that indicates which GPUs within that domain are physically capable of talking to each other. 
The partitioning of GPUs into a set of cliques is done at the NVSwitch layer, not at the individual node layer. All GPUs on a given node are guaranteed to have the same <ClusterUUID.Clique ID> pair. 

The ClusterUUID is a unique identifier for a given NVLink Domain. 
It can be queried on a GPU by GPU basis via the ``nvidia-smi`` commandline tool. 
All GPUs on a given node are guaranteed to have the same Cluster UUID. 

***************************************
About the ComputeDomain Custom Resource
***************************************

The NVIDIA GPU DRA Driver introduces a new custom resource called ComputeDomain, which creates a DRA ResourceClaimTemplate that you can reference in workloads. 
The ComputeDomain resources also creates a unique ResourceClaim for each worker that links it back to the ComputeDomain where the ResourceClaimTemplate is defined.

> If a subset of the nodes associated with a ComputeDomain are capable of communicating over IMEX, the NVIDIA Kubernetes DRA will set up a one-off IMEX domain to allow GPUs to communicate over their multi-node NVLink connections. Multiple IMEX domains will be created as necessary depending on the number (and availability) of nodes allocated to the ComputeDomain. 

Reviewer 4: Suggestion: IMEX -> Mulit-Node NVLink (MNNVL)

To create IMEX domains on your GPUs using the GPU DRA Driver, create a CustomDomain resource on your cluster. You can then reference the ResouceClaimTemplate in your workload specs as a ``resourceClaims.resourceClaimTemplateName``. 

.. literalinclude:: ./manifests/input/dra-compute-domain-crd.yaml
   :language: yaml
   :caption: Sample NVIDIA DRA Driver ComputeDomain Custom Resource Manifest

The following table describes some of the fields in the custom resource.

.. list-table::
   :header-rows: 1
   :widths: 20 60 20

   * - Field
     - Description
     - Default Value

   * - ``channel.resourceClaimTemplate.name`` (required)
     - Specifies the ``name`` of the ResourceClaimTemplate to create.
     - None

   * - ``numNodes`` (required)

>      - Specifies the number of nodes in the IMEX domain.

Reviewer 4: ```suggestion
     - Specifies the number of nodes in the Compute Domain.

```

Once scheduled to a node, the GPU DRA Driver adds a Node label for the ComputeDomain to the node where the workload has been scheduled to indicate the node is part of that ComputeDomain.
This label is used as a NodeSelector on the DaemonSet mentioned above to trigger the scheduling of its pods to specific nodes.

The addition of this Node label triggers the DaemonSet to deploy an IMEX daemon to that node and start running it.

When all daemons have been fully started, the GPU DRA Driver unblocks each worker, injects its IMEX channel into the worker and allows it to start running.

Once all workloads running in a ComputeDomain have run to completion, the label gets removed even if the ComputeDomain itself hasn't been deleted yet.
This allows these nodes to be reused for other ComputeDomains.

Node and Pod Affitnity Strategies
=================================

A ComputeDomain isn't strictly about IMEX channels—it's about running workloads across a group of compute nodes. 
This means even if some nodes are not IMEX capable, they can still be part of the same ComputeDomain.
You must apply NodeAffinity and PodAffinity rules to make sure your workloads run on IMEX capable nodes.

> For example you could set PodAffinity with a preferred topologyKey set to ``nvidia.com/gpu.clique`` for workloads to span multiple NVLink domains but want them packed as tightly as possible. Or use a required topologyKey set to ``nvidia.con/gpu.clique`` when you require all workloads deployed into the same NVLink domain, but don't care which one. 

Reviewer 5: I want to add my understanding here and then we can think about potentially rewording/reordering, simplifying.

Technically, when _requiring_ the value of  `nvidia.con/gpu.clique`  to be the same among all jobs then this is the same as saying: all jobs must land in the same domain&clique. And that means: all jobs are mutually reachable (all-to-all communication is possible). For a MNNVL setup this is what's typically wanted, and maybe we should lead with that.

So, I consider this use case to be more advanced or exotic:

>  for workloads to span multiple NVLink domains but want them packed as tightly as possible.

But maybe I don't know enough.

================================================

#. Add the NVIDIA Helm repository:

   .. code-block:: console

      $ helm repo add nvidia https://helm.ngc.nvidia.com/nvidia \
          && helm repo update

#. Install NVIDIA DRA Driver:

   .. code-block:: console

      $ helm install nvidia-dra-driver-gpu nvidia/nvidia-dra-driver-gpu \
            --version="25.3.0" \
            --create-namespace \
            --namespace nvidia-dra-driver-gpu \
            --set nvidiaDriverRoot=/run/nvidia/driver \
            --set nvidiaCtkPath=/usr/local/nvidia/toolkit/nvidia-ctk \
            --set resources.gpus.enabled=false

Common Chart Customization Options
==================================

The following options are available when using the Helm chart.
These options can be used with ``--set`` when installing with Helm.

The following table identifies the most frequently used options.
To view all the options, run ``helm show values nvidia/nvidia-dra-driver-gpu``.

.. list-table::
   :widths: 20 50 30
   :header-rows: 1

   * - Parameter
     - Description
     - Default

   * - ``nvidiaDriverRoot``
     - Specifies the driver root on the host.

>        For Operator-managed drivers (recommended), use ``/run/nvidia/driver``.

Reviewer 6: For clarity, lets be specific since our readers may not know what "operator" we are referring to.
```suggestion
       For GPU Operator-managed drivers (recommended), use ``/run/nvidia/driver``.

```

       For pre-installed drivers, use ``/``.
     - ``/``

   * - ``nvidiaCtkPath``
     - Specifies the path of The NVIDIA Container Toolkit CLI binary (nvidia-ctk) on the host.

>        For Operator-installed NVIDIA Container Toolkit (recommended), use ``/usr/local/nvidia/toolkit/nvidia-ctk``.

Reviewer 6: ```suggestion
       For GPU Operator-installed NVIDIA Container Toolkit (recommended), use ``/usr/local/nvidia/toolkit/nvidia-ctk``.

```

- A ComputeDomain resource. 

- A ResourceClaimTemplate to use for workload pods.

  This is used to request access to a unique IMEX channel on a set of compute nodes.
  Each worker of a multi-node job should reference this ResourceClaimTemplate to ensure that they have access to the same IMEX channel inside the appropriate ComputeDomain. 

- ResourceClaimTemplate for the DaemonSet.

  This is used to request access to a device that injects ComputeDomain-specific settings for an imex-daemon to run with (config.cfg, nodes_config.cfg, etc.). 
  Each worker of the DaemonSet mentioned below references this ResourceClaimTemplate to ensure they start up IMEX daemons associated with the appropriate ComputeDomain.

- A DaemonSet. 

  Depending on your workload requirements, this DaemonSet forms one (or more) IMEX domains by running IMEX daemons on the set of nodes in a ComputeDomain that requires them.
  When your workload is deployed, these daemons "follow" the workload pods to the nodes where they have been scheduled. 
  Through DRA, these daemons are guaranteed to be fully up and running before the workload pods that triggered their creation are allowed to run.

> As workload pods that reference a CustomDomain ResourceClaimTemplate, get scheduled they trigger the NVIDIA DRA Driver for GPUs to request access to the same IMEX channel on whatever node they land on. 

Reviewer 6: ```suggestion
As workload pods that reference a CustomDomain ResourceClaimTemplate get scheduled, they trigger the NVIDIA DRA Driver for GPUs to request access to the same IMEX channel on whatever node they land on. 

```

Node and Pod Affinity Strategies
=================================

A ComputeDomain isn't strictly about MNNVL—it's about running workloads across a group of compute nodes. 
This means even if some nodes are not MNNVL capable, they can still be part of the same ComputeDomain.
You must apply NodeAffinity and PodAffinity rules to make sure your workloads run on MNNVL capable nodes.

For example you could set PodAffinity with a required topologyKey set to ``nvidia.com/gpu.clique`` when you require all workloads deployed into the same NVLink domain, but don't care which one. 
Or use a preferred topologyKey set to ``nvidia.com/gpu.clique`` for workloads to span MNNVL domains but want them packed as tightly as possible. 

Create a CustomDomain and run a workload
========================================

The example manifest below, ``imex-channel-injection.yaml``, shows the creation of a CustomDomain, ``imex-channel-injection``, and a workload pod referencing the ResourceClaimTemplate, ``imex-channel-0``:

.. literalinclude:: ./manifests/input/imex-channel-injection.yaml
  :language: yaml

Apply the manifest.

.. code-block:: console

  $ kubectl apply -f imex-channel-injection.yaml

> #. Optional: View the ComputeDomain resource.

Reviewer 6: Shouldn't we also view the pod logs of the workload (which verifies that the imex channel device node was injected)?

Author: Added a step to view the pod logs

.. license-header
  SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  SPDX-License-Identifier: Apache-2.0

#############################################################
Multi-Node NVLink support with the NVIDIA DRA Driver for GPUs 
#############################################################

> The NVIDIA DRA Driver for GPUs is an additional component you can install alongside the GPU Operator that enables you to use the Kubernetes Dynamic Resource Allocation (DRA) feature to support Multi-Node NVLink in NVIDIA HGX GB200 NVL GPUs.

Author: from my understanding we are only supporting GB200, that what we have in the release notes

This page details more information about installing the DRA Driver for GPUs and examples of deploying workloads utilizing Multi-Node NVLink with NVIDIA HGX GB200 NVL systems.

> NVIDIA HGX GB200 NVL systems are designed specifically to leverage the use of IMEX channels to turn a rack of GPU machines, each with a small number of GPUs, into a giant supercomputer with up to 72 GPUs communicating at full NVLink bandwidth.

Reviewer 4: ```suggestion
NVIDIA HGX GB200 NVL systems are designed specifically to use Multi-Node NVLinks to turn a rack of GPU machines, each with a small number of GPUs, into a giant supercomputer with up to 72 GPUs communicating at full NVLink bandwidth.

```

IMEX channels are a low-level implementation detail.

> This allows you to get the most use out of your available GPUs without any additional latency burdens.

Reviewer 4: I don't think "burdens" is the right word here. "impact" maybe?

Reviewer 7: what about overhead?

Author: updated the wording here to be "overhead"

For more information about Kubernetes Dynamic Resource Allocation (DRA), refer to the `Kubernetes DRA documentation <https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/>`_.

************************************
About the NVIDIA DRA Driver for GPUs
************************************

> The NVIDIA DRA Driver for GPUs leverages the Kubernetes Dynamic Resource Allocation (DRA) API to support NVIDIA Multi-Node NVLink available in NVIDIA HGX GB200 NVL GPUs.

Reviewer 4: This seems to repeat what was said in the previou section. Was that the intent?

> The NVIDIA DRA Driver for GPUs introduces a Kubernetes custom resource named ComputeDomain where you can define your resource templates, and then reference the templates within your workload definitions. 

Reviewer 4: Should we "highlight" `ComputeDomain`?

> The NVIDIA DRA Driver for GPUs introduces a Kubernetes custom resource named ComputeDomain where you can define your resource templates, and then reference the templates within your workload definitions. 

Reviewer 4: The NVIDIA DRA Driver for GPUs introduced a Kubernetes custom resource named `ComputeDomain` which can be referenced in jobs that are expected to span multiple nodes. In the case of nodes conntected using MNNVL, the required resources are automatically provisioned to allow a set of GPUs to directly read and write each other's memory over a high-bandwidth NVLink.

> A ComputeDomain creates and manages an IMEX channel, a construct that allows a set of GPUs to directly read and write each other's memory over a high-bandwidth NVLink. 

Reviewer 4: Did we ever reach consensus on whether we want to actually mention "IMEX channels" here. (A compute domain can also be used on nodes that do not support MNNVL)

      [Mar 27 2025 15:47:10] [INFO] [tid 63] Started processing of incoming messages.
      [Mar 27 2025 15:47:10] [INFO] [tid 64] Started processing of incoming messages.
      [Mar 27 2025 15:47:10] [INFO] [tid 65] Started processing of incoming messages.
      [Mar 27 2025 15:47:10] [INFO] [tid 39] Creating gRPC channels to all peers (nPeers = 1).
      [Mar 27 2025 15:47:10] [INFO] [tid 66] Started processing of incoming messages.
      [Mar 27 2025 15:47:10] [INFO] [tid 39] IMEX_WAIT_FOR_QUORUM != FULL, continuing initialization without waiting for connections to all nodes.
      [Mar 27 2025 15:47:10] [INFO] [tid 67] Connection established to node 0 with ip address 10.115.131.8. Number of times connected: 1
      [Mar 27 2025 15:47:10] [INFO] [tid 39] GPU event successfully subscribed

  
#. Delete ``imex-channel-injection`` example.

   .. code-block:: console

      $ kubectl delete -f imex-channel-injection.yaml

   *Example Output*

   .. code-block:: output

      computedomain.resource.nvidia.com "imex-channel-injection" deleted
      pod "imex-channel-injection" deleted

******************************************************************

> Run a Multi-node nvbandwidth Test Requiring IMEX Channels with MPI

Author: @guptaNswati This is the start of the nvbandwith example

******************************************************************

This example demonstrates how to run a workload across multiple nodes using a ComputeDomain. 
The nvbandwidth test will measure the bandwidth between GPUs across different nodes using IMEX channels, helping you verify that your MNNVL setup is working correctly.

The following sample files use the following:

- `Kubeflow MPI Operator <https://www.kubeflow.org/docs/components/trainer/legacy-v1/user-guides/mpi/#installationr>`__  

> - A 2 node cluster with all nodes capable of running GPU workloads and meets [NVIDIA DRA Driver for GPUs prerequisites](#prerequisites)

Reviewer: @a-mccarthy We need to specify the number of GPUs. 
 
Please note that this example is for 2 node cluster which has 4 GPUs. Change the following according to your environment:
`spec.numNodes` | Total number of nodes in the cluster | `2` |
`spec.slotsPerWorker` | Number of GPUs per node, this should match the ppr number  | `4` |
`Worker.replicas` | Number of worker nodes | `2` |
`args` to mpirun | Adjust process-per-resource and total processes | `-ppr:4(numGPUs per node):node -np 8(numGPU x numNodes)` |

cc @ArangoGutierrez to help with formatting. Not sure whats the best way to specify these in detail. I have seen many people making mistake with these numbers so i think we should definitely add a note.

Author: @guptaNswati thanks! I've updated the docs to include the table, i also added some inline comments to callout how to update the example to a different cluster set up. can you take another look and make sure it looks ok?

Reviewer: LGTM. Thanks @a-mccarthy

.. license-header
  SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  SPDX-License-Identifier: Apache-2.0

#############################################################
Multi-Node NVLink support with the NVIDIA DRA Driver for GPUs 
#############################################################

The NVIDIA DRA Driver for GPUs is an additional component you can install alongside the GPU Operator that enables you to use the Kubernetes Dynamic Resource Allocation (DRA) feature to support Multi-Node NVLink in NVIDIA HGX GB200 NVL GPUs.
This page details more information about installing the DRA Driver for GPUs and examples of deploying workloads utilizing Multi-Node NVLink with NVIDIA HGX GB200 NVL systems.

> NVIDIA HGX GB200 NVL systems are designed specifically to leverage the use of IMEX channels to turn a rack of GPU machines, each with a small number of GPUs, into a giant supercomputer with up to 72 GPUs communicating at full NVLink bandwidth.

Reviewer 7: Should we expand IMEX once before we use this acronym everywhere else?

I believe IMEX stands for `Internode Memory Exchange`

For more information about Kubernetes Dynamic Resource Allocation (DRA), refer to the `Kubernetes DRA documentation <https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/>`_.

************************************
About the NVIDIA DRA Driver for GPUs
************************************

The NVIDIA DRA Driver for GPUs leverages the Kubernetes Dynamic Resource Allocation (DRA) API to support NVIDIA Multi-Node NVLink available in NVIDIA HGX GB200 NVL GPUs.
The NVIDIA DRA Driver for GPUs introduces a Kubernetes custom resource named ComputeDomain where you can define your resource templates, and then reference the templates within your workload definitions. 

A ComputeDomain creates and manages an IMEX channel, a construct that allows a set of GPUs to directly read and write each other's memory over a high-bandwidth NVLink. 
The NVLink connection may either be directly between GPUs on the same node or between GPUs on separate nodes connected by an NVSwitch. 
Once a ComputeDomain has been established for a set of GPUs, through an IMEX channel, the GPUs are free to read and write each other's memory via extensions to the CUDA memory call APIs.

> Kubernetes DRA, available as beta from Kubernetes v1.32, is an API for requesting and sharing resources between pods and containers inside a pod. 

Reviewer 7: ```suggestion
Kubernetes DRA, available as a beta feature from Kubernetes v1.32, is an API for requesting and sharing resources between pods and containers inside a pod. 

```

Reviewer 7: Thinking out loud: Should we expand DRA here?

Author: We have it written out earlier in the doc. But we can expand it out here, just for clarity

This feature treats specialized hardware as a definable and reusable object and provides the necessary primitives to support cross-node resources such as IMEX channels. 
A ComputeDomain uses DRA features to define IMEX channel resources that can be managed by Kubernetes.

Refer to the `Kubernetes DRA documentation`_ for details on this feature. 

**************************************
Install the NVIDIA DRA Driver for GPUs
**************************************

> The NVIDIA DRA Driver for GPUs is an additional component that can be installed alongside the GPU Operator on your Kubernetescluster.

Reviewer 7: ```suggestion
The NVIDIA DRA Driver for GPUs is an additional component that can be installed alongside the GPU Operator on your Kubernetes cluster.

```

Prerequisites
=============

> - A Mult-iNode NVIDIA HGX GB200 NVL GPUs with at least 2 GPUs Multi-Node NVLink support.

Reviewer 7: ```suggestion
- A Mult-Node NVIDIA HGX GB200 NVL GPUs with at least 2 GPUs Multi-Node NVLink support.

```

      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-bd6gn          1/1     Running   0          7m27s
      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-bzm6p          1/1     Running   0          7m26s
      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-xjm4p          1/1     Running   0          7m27s

#. Confirm that all GPU nodes are labeled with clique ids.
   The following command used `jq <https://jqlang.org/>`_ to format the output.

   .. code-block:: console

      $ (echo -e "NODE\tLABEL\tCLIQUE"; kubectl get nodes -o json | \
        jq -r '.items[] | [.metadata.name, "nvidia.com/gpu.clique", .metadata.labels["nvidia.com/gpu.clique"]] | @tsv') | \
        column -t

   *Example Output*

   .. code-block:: output
    
      NODE                  LABEL                  CLIQUE
      node1                 nvidia.com/gpu.clique  9277d399-0674-44a9-b64e-d85bb19ce2b0.32766
      node2                 nvidia.com/gpu.clique  9277d399-0674-44a9-b64e-d85bb19ce2b0.32766

The `NVIDIA GPU Feature Discovery <https://github.com/NVIDIA/k8s-device-plugin/tree/main/docs/gpu-feature-discovery>`_ adds a Clique ID to each GPU node. 

> This is an unique identifier within an NVLink domain (physically connected GPUs over NVLink) that indicates which GPUs within that domain are physically capable of talking to each other. 

Reviewer 7: ```suggestion
This is a unique identifier within an NVLink domain (physically connected GPUs over NVLink) that indicates which GPUs within that domain are physically capable of talking to each other. 

```

If a subset of the nodes associated with a ComputeDomain are capable of communicating over MNNVL, the NVIDIA DRA Driver for GPUs will set up a one-off IMEX domain to allow GPUs to communicate over their multi-node NVLink connections. Multiple IMEX domains will be created as necessary depending on the number and availability of nodes allocated to the ComputeDomain. 

You can then reference the ResourceClaimTemplate in your workload specs as a ``resourceClaims.resourceClaimTemplateName``. 

.. literalinclude:: ./manifests/input/dra-compute-domain-crd.yaml
   :language: yaml
   :caption: Sample NVIDIA DRA Driver ComputeDomain Custom Resource Manifest

The following table describes some of the fields in the custom resource.

.. list-table::
   :header-rows: 1
   :widths: 20 60 20

   * - Field
     - Description
     - Default Value

   * - ``channel.resourceClaimTemplate.name`` (required)
     - Specifies the ``name`` of the ResourceClaimTemplate to create.
     - None

   * - ``numNodes`` (required)
     - Specifies the number of nodes in the ComputeDomain.
     - None

> When you create a CustomDomain resource and configure a pod to reference it, the NVIDIA DRA Driver for GPUs will create the following resources on your cluster:

Reviewer 7: ```suggestion
When you create a ComputeDomain resource and configure a pod to reference it, the NVIDIA DRA Driver for GPUs will create the following resources on your cluster:
```

When all daemons have been fully started, the NVIDIA DRA Driver for GPUs unblocks each worker, injects its IMEX channel into the worker and allows it to start running.

Once all workloads running in a ComputeDomain have run to completion, the label gets removed even if the ComputeDomain itself hasn't been deleted yet.
This allows these nodes to be reused for other ComputeDomains.

Node and Pod Affinity Strategies
=================================

A ComputeDomain isn't strictly about MNNVL—it's about running workloads across a group of compute nodes. 
This means even if some nodes are not MNNVL capable, they can still be part of the same ComputeDomain.
You must apply NodeAffinity and PodAffinity rules to make sure your workloads run on MNNVL capable nodes.

For example you could set PodAffinity with a required topologyKey set to ``nvidia.com/gpu.clique`` when you require all workloads deployed into the same NVLink domain, but don't care which one. 
Or use a preferred topologyKey set to ``nvidia.com/gpu.clique`` for workloads to span MNNVL domains but want them packed as tightly as possible. 

> Create a CustomDomain and run a workload

Reviewer 7: ```suggestion
Create a ComputeDomain and run a workload
```

.. license-header
  SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  SPDX-License-Identifier: Apache-2.0

##########################
NVIDIA DRA Driver for GPUs
##########################

This page provides an overview of the NVIDIA DRA Driver for GPUs, its supported functionality, instructions for how to install it alongside the GPU Operator, and usage examples.

Before continuing, you should be familiar with the concept of Dynamic Resource Allocation (DRA) in Kubernetes (`docs <https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/>`_).

*******************
Supported Use Cases
*******************

The NVIDIA DRA Driver for GPUs v25.3.0 enables allocating

> * so-called ``ComputeDomain``\s for enabling Multi-Node NVLink (MNNVL) workloads on NVIDIA GB200 systems (full support).

Author: updated

      nvidia-dra-driver-k8s-dra-driver-kubelet-plugin-xjm4p          1/1     Running   0          7m27s

#. Confirm that all GPU nodes are labeled with clique ids.
   The following command used `jq <https://jqlang.org/>`_ to format the output.

   .. code-block:: console

      $ (echo -e "NODE\tLABEL\tCLIQUE"; kubectl get nodes -o json | \
        jq -r '.items[] | [.metadata.name, "nvidia.com/gpu.clique", .metadata.labels["nvidia.com/gpu.clique"]] | @tsv') | \
        column -t

   *Example Output*

   .. code-block:: output

      NODE                  LABEL                  CLIQUE
      node1                 nvidia.com/gpu.clique  9277d399-0674-44a9-b64e-d85bb19ce2b0.32766
      node2                 nvidia.com/gpu.clique  9277d399-0674-44a9-b64e-d85bb19ce2b0.32766

The `NVIDIA GPU Feature Discovery <https://github.com/NVIDIA/k8s-device-plugin/tree/main/docs/gpu-feature-discovery>`_ adds a Clique ID to each GPU node.
This is a unique identifier within an NVLink domain (physically connected GPUs over NVLink) that indicates which GPUs within that domain are physically capable of talking to each other.

> The partitioning of GPUs into a set of cliques is done at the NVSwitch layer, not at the individual node layer. All GPUs on a given node are guaranteed to have the same <ClusterUUID.Clique ID> pair.

Reviewer 8: And why is it a pair? Oh, does the pair refer to the ClusterUUID and the Clique ID? Would "value" be better. I think of "pair" as something like (X,Y). Later on, "<cluster-uuid, clique-id>" is referenced, so this was confusing for me.

Within the bracket notation, would it be better if "Clique ID" were "Clique_ID" to make it clear that it is one value?

******************************************************************

This example demonstrates how to run a workload across multiple nodes using a ComputeDomain.
The nvbandwidth test will measure the bandwidth between GPUs across different nodes using IMEX channels, helping you verify that your MNNVL setup is working correctly.

**Example notes:**

- This example uses `Kubeflow MPI Operator <https://www.kubeflow.org/docs/components/trainer/legacy-v1/user-guides/mpi/#installationr>`__.

- This example is configured for a 2 node cluster with 4 GPUs per node.

  If you are using a cluster with a different number of nodes and GPUs per node, you must adjust the following parameters in the sample files:

.. list-table::
   :widths: 15 55 30
   :header-rows: 1

   * - Parameter to update
     - Description
     - Value in example

   * - ``ComputeDomain.spec.numNodes``
     - Total number of nodes in the cluster
     - 2

   * - ``MPIJob.spec.slotsPerWorker``
     - Number of GPUs per node, this should match the ppr number
     - 4

   * - ``MPIJob.spec.mpiReplicaSpecs.Worker.replicas``
     - Number of worker nodes
     - 2

   * - ``mpirun`` command argument ``-ppr:4:node``

     -

>        * Number of GPUs per node as the process-per-resource number

Reviewer 8: Does this need to be a bullet item since there is only one?

This example demonstrates how to run a workload across multiple nodes using a ComputeDomain. 
The nvbandwidth test will measure the bandwidth between GPUs across different nodes using IMEX channels, helping you verify that your MNNVL setup is working correctly.

**Example notes:** 

- This example uses `Kubeflow MPI Operator <https://www.kubeflow.org/docs/components/trainer/legacy-v1/user-guides/mpi/#installationr>`__.

> - This example is configured for a 2 node cluster with 4 GPUS per node.

Reviewer 7: ```suggestion
- This example is configured for a 2 node cluster with 4 GPUs per node.

```

>   If you are using a cluster with different number of nodes and GPUs per node, you must adjust the following parameters in the sample files:

Reviewer 7: ```suggestion
  If you are using a cluster with a different number of nodes and GPUs per node, you must adjust the following parameters in the sample files:

```
